From 45798ae1d6005c0b3438b29f95ca07d2abb88cd3 Mon Sep 17 00:00:00 2001
From: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
Date: Tue, 24 Sep 2024 16:32:46 +0530
Subject: [PATCH] PENDING: allow logging domains to execute busybox_nosuid

Added policy module for busybox_nosuid and
also added policies allowing dmesg, syslogd, klogd to
execute busybox_nosuid.

Upstream-Status: Pending

Signed-off-by: Sasi Kumar Maddineni <quic_sasikuma@quicinc.com>
---
 policy/modules/admin/dmesg.te    |  2 +-
 policy/modules/system/busybox.fc |  3 +++
 policy/modules/system/busybox.if | 21 +++++++++++++++++++++
 policy/modules/system/busybox.te |  8 ++++++++
 policy/modules/system/logging.te |  4 +++-
 5 files changed, 36 insertions(+), 2 deletions(-)
 create mode 100644 policy/modules/system/busybox.fc
 create mode 100644 policy/modules/system/busybox.if
 create mode 100644 policy/modules/system/busybox.te

diff --git a/policy/modules/admin/dmesg.te b/policy/modules/admin/dmesg.te
index f1da315a9..afdcaaea3 100644
--- a/policy/modules/admin/dmesg.te
+++ b/policy/modules/admin/dmesg.te
@@ -36,7 +36,7 @@ fs_search_auto_mountpoints(dmesg_t)
 term_dontaudit_use_console(dmesg_t)
 
 domain_use_interactive_fds(dmesg_t)
-
+busybox_nosuid_execute(dmesg_t)
 files_list_etc(dmesg_t)
 files_read_etc_files(dmesg_t)
 files_read_usr_files(dmesg_t)
diff --git a/policy/modules/system/busybox.fc b/policy/modules/system/busybox.fc
new file mode 100644
index 000000000..28585bda0
--- /dev/null
+++ b/policy/modules/system/busybox.fc
@@ -0,0 +1,3 @@
+#Copyright (c) 2023 Qualcomm Innovation Center, Inc. All rights reserved.
+#SPDX-License-Identifier: BSD-3-Clause-Clear
+/usr/bin/busybox.nosuid         --      system_u:object_r:busybox_nosuid_exec_t:s0
diff --git a/policy/modules/system/busybox.if b/policy/modules/system/busybox.if
new file mode 100644
index 000000000..395d65bbc
--- /dev/null
+++ b/policy/modules/system/busybox.if
@@ -0,0 +1,21 @@
+#Copyright (c) 2023 Qualcomm Innovation Center, Inc. All rights reserved.
+#SPDX-License-Identifier: BSD-3-Clause-Clear
+## <summary>Busybox service</summary>
+
+######################################
+## <summary>
+##      Allow domain to execute busybox.nosuid
+## </summary>
+## <param name="domain">
+##      <summary>
+##      Domain allowed access.
+##      </summary>
+## </param>
+#
+interface(`busybox_nosuid_execute',`
+	gen_require(`
+		type busybox_nosuid_exec_t;
+	')
+
+	allow $1 busybox_nosuid_exec_t:file mmap_exec_file_perms;
+')
diff --git a/policy/modules/system/busybox.te b/policy/modules/system/busybox.te
new file mode 100644
index 000000000..094e6fc76
--- /dev/null
+++ b/policy/modules/system/busybox.te
@@ -0,0 +1,8 @@
+#Copyright (c) 2023 Qualcomm Innovation Center, Inc. All rights reserved.
+#SPDX-License-Identifier: BSD-3-Clause-Clear
+
+policy_module(busybox, 1.0.0)
+
+type busybox_nosuid_t;
+type busybox_nosuid_exec_t;
+init_system_domain(busybox_nosuid_t, busybox_nosuid_exec_t)
diff --git a/policy/modules/system/logging.te b/policy/modules/system/logging.te
index eea78ffc5..291150cee 100644
--- a/policy/modules/system/logging.te
+++ b/policy/modules/system/logging.te
@@ -96,7 +96,9 @@ files_type(syslogd_var_lib_t)
 type var_log_t;
 logging_log_file(var_log_t)
 files_mountpoint(var_log_t)
-
+busybox_nosuid_execute(klogd_t)
+busybox_nosuid_execute(syslogd_t)
+logging_read_syslog_config(syslogd_t)
 ifdef(`enable_mls',`
 	init_ranged_daemon_domain(auditd_t, auditd_exec_t, mls_systemhigh)
 	init_ranged_daemon_domain(syslogd_t, syslogd_exec_t, mls_systemhigh)
-- 
2.25.1
